### ISSUE TITLE
Bug: userland rules depending on `@typescript-eslint/utils` will fail compilation if `exactOptionalPropertyTypes` is set

### ISSUE THREAD

**Issue by jakehamtexas** [typescript-eslint/typescript-eslint#6307]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
// Importing `TSESTree` for the use of `AST_NODE_TYPES` can reproduce the error given the proper `tsconfig`.
import { TSESTree } from "@typescript-eslint/utils";
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "exactOptionalPropertyTypes": true
  }
}
```

### Expected Result

Since this is a so-called "Recommended" compiler option, I would expect my rule project code to compile successfully with any dependency from `@typescript-eslint`.

### Actual Result

![image](https://user-images.githubusercontent.com/31600776/211130082-c797e80a-6764-4028-ad3e-418a3d16539e.png)

### Additional Info

This looks like it's fairly intentional (or at least unavoidable) given the comment in `ClassBase`, here: 
![image](https://user-images.githubusercontent.com/31600776/211130506-1e37a96e-7ba6-450e-9380-f8939d638333.png)

Hopefully, it is possible to use a generic type parameter for the interface `ClassBase` in the following manner, although given that the types are generated, I don't know how feasible it would be to implement a fix including generic type parameters:

```typescript
// The default (`boolean`) for the generic parameter can be useful for backwards compatibility
declare interface ClassBase<TAbstract extends boolean | undefined = boolean> extends BaseNode {
    abstract?: TAbstract;

    // etc.
}

// elsewhere
export declare interface ClassExpression extends ClassBase<undefined> {
    type: AST_NODE_TYPES.ClassExpression;
    // This member can simply be elided now
    // abstract?: undefined;
    declare?: undefined;
    decorators?: undefined;
}
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/utils`             | `5.48.0` |
| `TypeScript`                           | `4.9.4` |

**Comment by bradzacher** [comment]
I would suggest using [`skipLibCheck`] (https://www.typescriptlang.org/tsconfig/#skipLibCheck) for now so TS won't attempt to validate our types using different compiler options to what they were designed with.

We'll look into this and determine if we want to turn this on for our project.

---

> Since this is a so-called "Recommended" compiler option

Where did you see it recommended? It's not part of the `strict` compiler option, so it's not super recommended by the TS team at least.

**Comment by Josh-Cena** [comment]
@bradzacher It's recommended here: https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes

It's not in `strict`, because that would break too much user code. It's just an engineering concern, not a prescriptive stance.

**Comment by jakehamtexas** [comment]
> I would suggest using [skipLibCheck](https://www.typescriptlang.org/tsconfig/#skipLibCheck) for now so TS won't attempt to validate our types using different compiler options to what they were designed with.

Thank you, I forgot that this exists, and I'm kind of surprised that I forgot to include it in my own `tsconfig.json`.

> We'll look into this and determine if we want to turn this on for our project.

As `skipLibCheck` effectively removes the original concern and practically, this issue is now just reporting on a lack of adoption of the `exactOptionalPropertyTypes` compiler option in `@typescript-eslint/utils` and its depending packages, I'm not sure that there's much need to action on anything here. At the least it might be rebranded as a repo improvement, instead of a bug?

**Comment by JoshuaKGoldberg** [comment]
Yeah it'd be nice to investigate this internally. Not sure yet whether it's something we particularly want, but I'd guess so.

Let's wait until v6's #5886 goes in, since it reworks a bunch of types internally.

**Comment by RebeccaStevens** [comment]
> I would suggest using [`skipLibCheck`](https://www.typescriptlang.org/tsconfig/#skipLibCheck) for now so TS won't attempt to validate our types using different compiler options to what they were designed with.

I use the follow command to run type checking when I have libraries that don't work with `exactOptionalPropertyTypes`.

```sh
tsc --noEmit --skipLibCheck && tsc --noEmit --exactOptionalPropertyTypes false
```

That way, I'm still checking my code with `exactOptionalPropertyTypes` but allowing libraries not to need it.

**Comment by voxpelli** [comment]
Just a note:

`skipLibCheck` means "[Skip type checking of **declaration files**](https://www.typescriptlang.org/tsconfig#skipLibCheck)" no matter if those `.d.ts` files live within `node_modules/` or in your own project.

So, adding `skipLibCheck` is really only a feasible solution when you have no manually written `.d.ts`. As soon as you have then its no good.

See eg. this issue where we were a few who realized this: https://github.com/sindresorhus/tsconfig/issues/15

**Comment by JoshuaKGoldberg** [comment]
v6 was merged a bit back, so this is now accepting PRs!

Note that per https://github.com/typescript-eslint/typescript-eslint/issues/6307#issuecomment-1374564399, it's not 100% clear what enabling `exactOptionalPropertyTypes` would cause. It might be straightforward. It might be super duper convoluted. Implementer beware. 😄

**Comment by jakehamtexas** [comment]
I'm going to give it a shot over the next week or two. Hopefully, nobody else beats me to it :)

**Comment by unrevised6419** [comment]
@jkhmtx did you have time to look into it? 👀
